File: app/src/main/java/kfzkennzeichen/gisbertamm/de/kfz_kennzeichen/persistence/DatabaseHandler.java
package kfzkennzeichen.gisbertamm.de.kfz_kennzeichen.persistence;

import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteException;
import android.database.sqlite.SQLiteOpenHelper;
import android.util.Log;

import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;


public class DatabaseHandler extends SQLiteOpenHelper {
    private static final int DATABASE_VERSION = 3;

    private static final String DATABASE_NAME = "NumberplateCodesManager.sqlite";

    private static final String TABLE_NUMBERPLATE_CODES = "numberplate_codes";
    private static final String TABLE_JOKES = "jokes";

    private static final String COLUMN_ID = "_id";
    private static final String COLUMN_CODE = "code";
    private static final String COLUMN_DISTRICT = "district";
    private static final String COLUMN_DISTRICT_CENTER = "district_center";
    private static final String COLUMN_STATE = "state";
    private static final String COLUMN_DISTRICT_WIKIPEDIA_URL = "district_wikipedia_url";
    public static final String WIKIPEDIA_BASE_URL = "https://de.wikipedia.org";
    private static final List<String[]> data = new ArrayList<String[]>();

    //The Android's default system path of your application database.
    private static String DB_PATH = "/data/data/kfzkennzeichen.gisbertamm.de.kfz_kennzeichen/databases/";

    private SQLiteDatabase dataBase;

    private final Context context;

    /**
     * Constructor
     * Takes and keeps a reference of the passed context in order to access to the application assets and resources.
     *
     * @param context
     */
    public DatabaseHandler(Context context) {
        super(context, DATABASE_NAME, null, DATABASE_VERSION);
        this.context = context;
        // copy database implicitly when class instance is created for the first time
        createDataBase();
    }

    /**
     * Creates a empty database on the system and rewrites it with your own database.
     */
    public void createDataBase() {

        SQLiteDatabase dbExist = checkDataBase();

        if (dbExist != null) {
            int version = dbExist.getVersion();
            Log.d(this.getClass().getSimpleName(), "EXISTING DATABASE VERSION: " + version);
            dbExist.close();

            if (DATABASE_VERSION > version) {
                // update existing database by overriding it
                copyDataBase();
            }
        } else {

            //By calling this method an empty database will be created into the default system path
            //of your application so we are gonna be able to overwrite that database with our database.
            this.getReadableDatabase();

            copyDataBase();

        }
    }

    /**
     * Check if the database already exist to avoid re-copying the file each time you open the application.
     *
     * @return true if it exists, false if it doesn't
     */
    private SQLiteDatabase checkDataBase() {

        SQLiteDatabase checkDB = null;

        try {
            String myPath = DB_PATH + DATABASE_NAME;
            checkDB = SQLiteDatabase.openDatabase(myPath, null, SQLiteDatabase.OPEN_READONLY);

        } catch (SQLiteException e) {

            //database does't exist yet.

        }
        return checkDB;
    }

    /**
     * Copies your database from your local assets-folder to the just created empty database in the
     * system folder, from where it can be accessed and handled.
     * This is done by transfering bytestream.
     */
    private void copyDataBase() {
        try {
            //Open your local db as the input stream
            InputStream myInput = context.getAssets().open(DATABASE_NAME);

            // Path to the just created empty db
            String outFileName = DB_PATH + DATABASE_NAME;

            Log.d(this.getClass().getSimpleName(),
                    "Copying database from assets/ " + DATABASE_NAME + " to " + outFileName);

            //Open the empty db as the output stream
            OutputStream myOutput = new FileOutputStream(outFileName);

            //transfer bytes from the inputfile to the outputfile
            byte[] buffer = new byte[1024];
            int length;
            while ((length = myInput.read(buffer)) > 0) {
                myOutput.write(buffer, 0, length);
            }

            //Close the streams
            myOutput.flush();
            myOutput.close();
            myInput.close();
        } catch (IOException e) {
            throw new RuntimeException("Cannot copy database", e);
        }

    }

    public void openDataBase() throws SQLException {

        //Open the database
        String myPath = DB_PATH + DATABASE_NAME;
        dataBase = SQLiteDatabase.openDatabase(myPath, null, SQLiteDatabase.OPEN_READONLY);

    }

    @Override
    public synchronized void close() {

        if (dataBase != null)
            dataBase.close();

        super.close();

    }

    @Override
    public void onCreate(SQLiteDatabase db) {
        // do nothing because database is provided from assets
    }

    @Override
    public void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion) {
        // do nothing because database is provided from assets
    }

    public SavedEntry searchForCode(String code) {
        try {
            this.openDataBase();
        } catch (SQLException e) {
            e.printStackTrace();
            return null;
        }

        Cursor cursor = this.dataBase.query(TABLE_NUMBERPLATE_CODES, new String[]{"*"}, COLUMN_CODE + "=?",
                new String[]{String.valueOf(code)}, null, null, null, null);

        if (cursor != null) {
            cursor.moveToFirst();
        } else {
            Log.e(this.getClass().getSimpleName(), "cursor is null");
            return null;
        }

        SavedEntry savedEntry = null;

        if (cursor.getCount() > 0) {
            savedEntry = new SavedEntry(Integer.parseInt(cursor.getString(0)),
                    cursor.getString(1), cursor.getString(2), cursor.getString(3),
                    cursor.getString(4), cursor.getString(5));

            if (!addJokes(code, savedEntry)) return null;
        }

        return savedEntry;
    }

    private boolean addJokes(String code, SavedEntry savedEntry) {
        Cursor cursor2 = this.dataBase.query(TABLE_JOKES, new String[]{"*"}, COLUMN_CODE + "=?",
                new String[]{String.valueOf(code)}, null, null, null, null);

        if (cursor2 != null) {
            cursor2.moveToFirst();
        } else {
            Log.e(this.getClass().getSimpleName(), "cursor2 is null");
            return false;
        }
        if (cursor2.getCount() > 0) {
            while (!cursor2.isAfterLast()) {
                final String joke = cursor2.getString(2);
                savedEntry.setJoke(joke);
                cursor2.moveToNext();
            }
        }
        return true;
    }

    public SavedEntry searchRandom() {
        try {
            this.openDataBase();
        } catch (SQLException e) {
            e.printStackTrace();
            return null;
        }

        Cursor cursor = this.dataBase.query(TABLE_NUMBERPLATE_CODES + " ORDER BY RANDOM() LIMIT 1",
                new String[]{"*"}, null, null, null, null, null);
        if (cursor != null) {
            cursor.moveToFirst();
        } else {
            Log.e(this.getClass().getSimpleName(), "cursor is null");
            return null;
        }

        SavedEntry savedEntry = null;

        if (cursor.getCount() > 0) {
            savedEntry = new SavedEntry(Integer.parseInt(cursor.getString(0)),
                    cursor.getString(1), cursor.getString(2), cursor.getString(3),
                    cursor.getString(4), cursor.getString(5));
        }

        if (!addJokes(cursor.getString(1), savedEntry)) return null;

        return savedEntry;
    }
}

File: app/src/main/java/kfzkennzeichen/gisbertamm/de/kfz_kennzeichen/persistence/DatabaseHandler.kt
package kfzkennzeichen.gisbertamm.de.kfz_kennzeichen.persistence

import android.content.Context
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteException
import android.database.sqlite.SQLiteOpenHelper
import android.util.Log
import java.io.FileOutputStream
import java.io.IOException
import java.io.OutputStream
import java.sql.SQLException
import java.util.*

class DatabaseHandler(private val context: Context) : SQLiteOpenHelper(context, DATABASE_NAME, null, DATABASE_VERSION) {
    private var dataBase: SQLiteDatabase? = null

    /**
     * Creates a empty database on the system and rewrites it with your own database.
     */
    fun createDataBase() {
        val dbExist = checkDataBase()
        if (dbExist != null) {
            val version = dbExist.version
            Log.d(this.javaClass.simpleName, "EXISTING DATABASE VERSION: $version")
            dbExist.close()
            if (DATABASE_VERSION > version) {
                // update existing database by overriding it
                copyDataBase()
            }
        } else {

            //By calling this method an empty database will be created into the default system path
            //of your application so we are gonna be able to overwrite that database with our database.
            this.readableDatabase
            copyDataBase()
        }
    }

    /**
     * Check if the database already exist to avoid re-copying the file each time you open the application.
     *
     * @return true if it exists, false if it doesn't
     */
    private fun checkDataBase(): SQLiteDatabase? {
        var checkDB: SQLiteDatabase? = null
        try {
            checkDB = context.openOrCreateDatabase(DATABASE_NAME, Context.MODE_PRIVATE, null)
        } catch (e: SQLiteException) {

            //database does't exist yet.
        }
        return checkDB
    }

    /**
     * Copies your database from your local assets-folder to the just created empty database in the
     * system folder, from where it can be accessed and handled.
     * This is done by transfering bytestream.
     */
    private fun copyDataBase() {
        try {
            //Open your local db as the input stream
            val myInput = context.assets.open(DATABASE_NAME)

            // Path to the just created empty db
            val outFileName = context.getDatabasePath(DATABASE_NAME).path
            Log.d(this.javaClass.simpleName,
                    "Copying database from assets/ " + DATABASE_NAME + " to " + outFileName)

            //Open the empty db as the output stream
            val myOutput: OutputStream = FileOutputStream(outFileName)

            //transfer bytes from the inputfile to the outputfile
            val buffer = ByteArray(1024)
            var length: Int
            while (myInput.read(buffer).also { length = it } > 0) {
                myOutput.write(buffer, 0, length)
            }

            //Close the streams
            myOutput.flush()
            myOutput.close()
            myInput.close()
        } catch (e: IOException) {
            throw RuntimeException("Cannot copy database", e)
        }
    }

    @Throws(SQLException::class)
    fun openDataBase() {

        //Open the database
        dataBase = context.openOrCreateDatabase(DATABASE_NAME, Context.MODE_PRIVATE, null)
    }

    @Synchronized
    override fun close() {
        if (dataBase != null) dataBase!!.close()
        super.close()
    }

    override fun onCreate(db: SQLiteDatabase) {
        // do nothing because database is provided from assets
    }

    override fun onUpgrade(db: SQLiteDatabase, oldVersion: Int, newVersion: Int) {
        // do nothing because database is provided from assets
    }

    fun searchForCode(code: String): SavedEntry? {
        try {
            openDataBase()
        } catch (e: SQLException) {
            e.printStackTrace()
            return null
        }
        val cursor = dataBase!!.query(TABLE_NUMBERPLATE_CODES, arrayOf("*"), COLUMN_CODE + "=?", arrayOf(code), null, null, null, null)
        if (cursor != null) {
            cursor.moveToFirst()
        } else {
            Log.e(this.javaClass.simpleName, "cursor is null")
            return null
        }
        var savedEntry: SavedEntry? = null
        if (cursor.count > 0) {
            savedEntry = SavedEntry(cursor.getString(0).toInt(),
                    cursor.getString(1), cursor.getString(2), cursor.getString(3),
                    cursor.getString(4), cursor.getString(5))
            if (!addJokes(code, savedEntry)) return null
        }
        return savedEntry
    }

    private fun addJokes(code: String, savedEntry: SavedEntry?): Boolean {
        val cursor = dataBase!!.query(TABLE_JOKES, arrayOf("*"), COLUMN_CODE + "=?", arrayOf(code), null, null, null, null)
        if (cursor != null) {
            cursor.moveToFirst()
        } else {
            Log.e(this.javaClass.simpleName, "cursor is null")
            return false
        }
        if (cursor.count > 0) {
            while (!cursor.isAfterLast) {
                val joke = cursor.getString(2)
                savedEntry!!.setJoke(joke)
                cursor.moveToNext()
            }
        }
        return true
    }

    fun searchRandom(): SavedEntry? {
        try {
            openDataBase()
        } catch (e: SQLException) {
            e.printStackTrace()
            return null
        }
        val cursor = dataBase!!.query(TABLE_NUMBERPLATE_CODES + " ORDER BY RANDOM() LIMIT 1", arrayOf("*"), null, null, null, null, null)
        if (cursor != null) {
            cursor.moveToFirst()
        } else {
            Log.e(this.javaClass.simpleName, "cursor is null")
            return null
        }
        var savedEntry: SavedEntry? = null
        if (cursor.count > 0) {
            savedEntry = SavedEntry(cursor.getString(0).toInt(),
                    cursor.getString(1), cursor.getString(2), cursor.getString(3),
                    cursor.getString(4), cursor.getString(5))
        }
        return if (!addJokes(cursor.getString(1), savedEntry)) null else savedEntry
    }

    fun createStatistics() {
        try {
            openDataBase()
        } catch (e: SQLException) {
            e.printStackTrace()
        }
        val cursor = dataBase!!.query(TABLE_JOKES, arrayOf("COUNT(*)"), "$COLUMN_JOKES like ?", arrayOf("%Esel%"), null, null, null, null)
        cursor?.moveToFirst() ?: Log.e(this.javaClass.simpleName, "cursor is null")
        if (cursor!!.count > 0) {
            while (!cursor.isAfterLast) {
                val code = cursor.getString(0)
                Log.d(TAG, "Number of Esel is $code")
                cursor.moveToNext()
            }
        }
    }

    companion object {
        const val WIKIPEDIA_BASE_URL = "https://de.wikipedia.org"
        const val TAG = "DatabaseHandler"
        private const val DATABASE_VERSION = 9
        private const val DATABASE_NAME = "NumberplateCodesManager.sqlite"
        private const val TABLE_NUMBERPLATE_CODES = "numberplate_codes"
        private const val TABLE_JOKES = "jokes"
        private const val COLUMN_JOKES = "jokes"
        private const val COLUMN_ID = "_id"
        private const val COLUMN_CODE = "code"
        private const val COLUMN_DISTRICT = "district"
        private const val COLUMN_DISTRICT_CENTER = "district_center"
        private const val COLUMN_STATE = "state"
        private const val COLUMN_DISTRICT_WIKIPEDIA_URL = "district_wikipedia_url"
        private val data: List<Array<String>> = ArrayList()
    }

    /**
     * Constructor
     * Takes and keeps a reference of the passed context in order to access to the application assets and resources.
     *
     * @param context
     */
    init {
        // copy database implicitly when class instance is created for the first time
        createDataBase()
    }
}